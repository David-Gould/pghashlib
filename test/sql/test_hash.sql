
select hash_string('', 'crc32');
select hash_string('a', 'crc32');
select hash_string('abcdefg', 'crc32');
select hash_string('defg', 'crc32', hash_string('abc', 'crc32'));
select encode(hash128_string('abcdefg', 'crc32'), 'hex');

select hash_string('', 'lookup2');
select hash_string('a', 'lookup2');
select hash_string('abcdefg', 'lookup2');
select encode(hash128_string('abcdefg', 'lookup2'), 'hex');
select hash_string('01234567890', 'lookup2');
select hash_string('012345678901', 'lookup2');
select hash_string('0123456789012', 'lookup2');
select hash_string('01234567890123', 'lookup2');

select hash_string('', 'lookup3');
select hash_string('a', 'lookup3');
select hash_string('abcdefg', 'lookup3');
select hash_string('01234567890', 'lookup2');
select hash_string('012345678901', 'lookup2');
select hash_string('0123456789012', 'lookup2');
select hash_string('01234567890123', 'lookup2');

select hash_string('', 'lookup3be');
select hash_string('a', 'lookup3be');
select hash_string('abcdefg', 'lookup3be');
select encode(hash128_string('abcdefg', 'lookup3be'), 'hex');
select hash_string('01234567890', 'lookup2');
select hash_string('012345678901', 'lookup2');
select hash_string('0123456789012', 'lookup2');
select hash_string('01234567890123', 'lookup2');

select hash_string('', 'murmur3');
select hash_string('a', 'murmur3');
select hash_string('abcdefg', 'murmur3');
select encode(hash128_string('abcdefg', 'murmur3'), 'hex');

select hash_string('', 'pgsql84');
select hash_string('a', 'pgsql84');
select hash_string('abcdefg', 'pgsql84');
select encode(hash128_string('abcdefg', 'pgsql84'), 'hex');
select hash_string('01234567890', 'lookup2');
select hash_string('012345678901', 'lookup2');
select hash_string('0123456789012', 'lookup2');
select hash_string('01234567890123', 'lookup2');

select hash64_string('', 'lookup3le');
select hash64_string('a', 'lookup3le');
select hash64_string('abcdefg', 'lookup3le');
select encode(hash128_string('abcdefg', 'lookup3le'), 'hex');
select hash64_string('01234567890', 'lookup2');
select hash64_string('012345678901', 'lookup2');
select hash64_string('0123456789012', 'lookup2');
select hash64_string('01234567890123', 'lookup2');

select hash64_string('', 'city64');
select hash64_string('a', 'city64');
select hash64_string('abcdefg', 'city64');
select encode(hash128_string('abcdefg', 'city64'), 'hex');
select hash64_string('0123456789abcdef', 'city64');
select hash64_string('0123456789abcdef0', 'city64');
select hash64_string('0123456789abcdef0123456789abcdef', 'city64');
select hash64_string('0123456789abcdef0123456789abcdef0', 'city64');
select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcde', 'city64');
select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef', 'city64');
select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0', 'city64');


select hash64_string('', 'city128');
select hash64_string('a', 'city128');
select hash64_string('abcdefg', 'city128');
select encode(hash128_string('abcdefg', 'city128'), 'hex');
select hash64_string('0123456789abcdef', 'city128');
select hash64_string('0123456789abcdef0', 'city128');
select hash64_string('0123456789abcdef0123456789abcdef', 'city128');
select hash64_string('0123456789abcdef0123456789abcdef0', 'city128');
select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcde', 'city128');
select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef', 'city128');
select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0', 'city128');

select hash64_string('', 'spooky');
select hash64_string('a', 'spooky');
select hash64_string('abcdefg', 'spooky');
select encode(hash128_string('abcdefg', 'spooky'), 'hex');
select hash64_string('0123456789abcdef', 'spooky');
select hash64_string('0123456789abcdef0', 'spooky');
select hash64_string('0123456789abcdef0123456789abcdef', 'spooky');
select hash64_string('0123456789abcdef0123456789abcdef0', 'spooky');
select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcde', 'spooky');
select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef', 'spooky');
select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0', 'spooky');

SELECT encode(hash128_string('', 'md5'), 'hex');
-- d41d8cd98f00b204e9800998ecf8427e
SELECT encode(hash128_string('a', 'md5'), 'hex');
-- 0cc175b9c0f1b6a831c399e269772661
SELECT encode(hash128_string('abc', 'md5'), 'hex');
-- 900150983cd24fb0d6963f7d28e17f72
SELECT encode(hash128_string('message digest', 'md5'), 'hex');
-- f96b697d7cb7938d525a2f31aaf161d0
SELECT encode(hash128_string('abcdefghijklmnopqrstuvwxyz', 'md5'), 'hex');
-- c3fcd3d76192e4007dfb496cca67e13b
SELECT encode(hash128_string('ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789', 'md5'), 'hex');
-- d174ab98d277d9f5a5611c2c9f419d9f
SELECT encode(hash128_string('12345678901234567890123456789012345678901234567890123456789012345678901234567890', 'md5'), 'hex');
-- 57edf4a22be3c955ac49da2e2107b67a

--
-- integer hashes
--

select hash_int4(0, 'jenkins');
select hash_int4(12345678, 'jenkins');
select hash_int4(1234567890123456789::int8, 'jenkins');

select hash_int4(0, 'wang32');
select hash_int4(12345678, 'wang32');
select hash_int4(1234567890123456789::int8, 'wang32');

select hash_int4(0, 'wang32mult');
select hash_int4(12345678, 'wang32mult');
select hash_int4(1234567890123456789::int8, 'wang32mult');

select hash_int8(0, 'wang64');
select hash_int8(12345678, 'wang64');
select hash_int8(1234567890123456789::int8, 'wang64');

select hash_int8(0, 'wang64to32');
select hash_int8(12345678, 'wang64to32');
select hash_int8(1234567890123456789::int8, 'wang64to32');

-- make sure conversion to int4 works correctly
select count(hash_int8(x, 'wang64to32')::int4) from generate_series(1,1000) x;


select x, hash_int4(x + 6, 'jenkins') from generate_series(1, 10) x order by hash_int4(x + 6, 'jenkins');


