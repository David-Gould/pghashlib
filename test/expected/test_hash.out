select hash_string('', 'crc32');
 hash_string 
-------------
           0
(1 row)

select hash_string('a', 'crc32');
 hash_string 
-------------
  -390611389
(1 row)

select hash_string('abcdefg', 'crc32');
 hash_string 
-------------
   824863398
(1 row)

select hash_string('defg', 'crc32', hash_string('abc', 'crc32'));
 hash_string 
-------------
   824863398
(1 row)

select encode(hash128_string('abcdefg', 'crc32'), 'hex');
              encode              
----------------------------------
 a66a2a31000000000000000000000000
(1 row)

select hash_string('', 'lookup2');
 hash_string 
-------------
  -863449762
(1 row)

select hash_string('a', 'lookup2');
 hash_string 
-------------
   542836691
(1 row)

select hash_string('abcdefg', 'lookup2');
 hash_string 
-------------
 -1528042926
(1 row)

select encode(hash128_string('abcdefg', 'lookup2'), 'hex');
              encode              
----------------------------------
 2c76e6b9ef60e9610000000000000000
(1 row)

select hash_string('01234567890', 'lookup2');
 hash_string 
-------------
  -365883078
(1 row)

select hash_string('012345678901', 'lookup2');
 hash_string 
-------------
   606216542
(1 row)

select hash_string('0123456789012', 'lookup2');
 hash_string 
-------------
  -111811398
(1 row)

select hash_string('01234567890123', 'lookup2');
 hash_string 
-------------
   609638926
(1 row)

select hash_string('', 'lookup3');
 hash_string 
-------------
  -559038737
(1 row)

select hash_string('a', 'lookup3');
 hash_string 
-------------
  1490454280
(1 row)

select hash_string('abcdefg', 'lookup3');
 hash_string 
-------------
 -1323641691
(1 row)

select hash_string('01234567890', 'lookup2');
 hash_string 
-------------
  -365883078
(1 row)

select hash_string('012345678901', 'lookup2');
 hash_string 
-------------
   606216542
(1 row)

select hash_string('0123456789012', 'lookup2');
 hash_string 
-------------
  -111811398
(1 row)

select hash_string('01234567890123', 'lookup2');
 hash_string 
-------------
   609638926
(1 row)

select hash_string('', 'lookup3be');
 hash_string 
-------------
  -559038737
(1 row)

select hash_string('a', 'lookup3be');
 hash_string 
-------------
  -454251968
(1 row)

select hash_string('abcdefg', 'lookup3be');
 hash_string 
-------------
 -1279733517
(1 row)

select encode(hash128_string('abcdefg', 'lookup3be'), 'hex');
              encode              
----------------------------------
 f3d0b8b3ab7b723f0000000000000000
(1 row)

select hash_string('01234567890', 'lookup2');
 hash_string 
-------------
  -365883078
(1 row)

select hash_string('012345678901', 'lookup2');
 hash_string 
-------------
   606216542
(1 row)

select hash_string('0123456789012', 'lookup2');
 hash_string 
-------------
  -111811398
(1 row)

select hash_string('01234567890123', 'lookup2');
 hash_string 
-------------
   609638926
(1 row)

select hash_string('', 'murmur3');
 hash_string 
-------------
           0
(1 row)

select hash_string('a', 'murmur3');
 hash_string 
-------------
  1009084850
(1 row)

select hash_string('abcdefg', 'murmur3');
 hash_string 
-------------
 -2009294074
(1 row)

select encode(hash128_string('abcdefg', 'murmur3'), 'hex');
              encode              
----------------------------------
 069b3c88000000000000000000000000
(1 row)

select hash_string('', 'pgsql84');
 hash_string 
-------------
 -1477818771
(1 row)

select hash_string('a', 'pgsql84');
 hash_string 
-------------
  1075015857
(1 row)

select hash_string('abcdefg', 'pgsql84');
 hash_string 
-------------
   501636814
(1 row)

select encode(hash128_string('abcdefg', 'pgsql84'), 'hex');
              encode              
----------------------------------
 ce5ee61dcf458aaa0000000000000000
(1 row)

select hash_string('01234567890', 'lookup2');
 hash_string 
-------------
  -365883078
(1 row)

select hash_string('012345678901', 'lookup2');
 hash_string 
-------------
   606216542
(1 row)

select hash_string('0123456789012', 'lookup2');
 hash_string 
-------------
  -111811398
(1 row)

select hash_string('01234567890123', 'lookup2');
 hash_string 
-------------
   609638926
(1 row)

select hash64_string('', 'lookup3le');
    hash64_string     
----------------------
 -2401053088876216593
(1 row)

select hash64_string('a', 'lookup3le');
    hash64_string    
---------------------
 6351843130003064584
(1 row)

select hash64_string('abcdefg', 'lookup3le');
    hash64_string    
---------------------
 6461572128488215717
(1 row)

select encode(hash128_string('abcdefg', 'lookup3le'), 'hex');
              encode              
----------------------------------
 a5d41ab19d1dac590000000000000000
(1 row)

select hash64_string('01234567890', 'lookup2');
    hash64_string     
----------------------
 -4314575613073485510
(1 row)

select hash64_string('012345678901', 'lookup2');
    hash64_string     
----------------------
 -2387840458343440034
(1 row)

select hash64_string('0123456789012', 'lookup2');
    hash64_string    
---------------------
 4176792262876193978
(1 row)

select hash64_string('01234567890123', 'lookup2');
    hash64_string    
---------------------
 3256775724372482574
(1 row)

select hash64_string('', 'city64');
    hash64_string     
----------------------
 -7286425919675154353
(1 row)

select hash64_string('a', 'city64');
    hash64_string    
---------------------
 2603192927274642682
(1 row)

select hash64_string('abcdefg', 'city64');
    hash64_string     
----------------------
 -9081257160914143494
(1 row)

select encode(hash128_string('abcdefg', 'city64'), 'hex');
              encode              
----------------------------------
 fa76bd09bae4f8810000000000000000
(1 row)

select hash64_string('0123456789abcdef', 'city64');
   hash64_string    
--------------------
 692747204941329175
(1 row)

select hash64_string('0123456789abcdef0', 'city64');
    hash64_string    
---------------------
 5496319037578566379
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef', 'city64');
    hash64_string    
---------------------
 1759797222395115254
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0', 'city64');
    hash64_string     
----------------------
 -6142652710596976713
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcde', 'city64');
    hash64_string     
----------------------
 -8539013451965526413
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef', 'city64');
    hash64_string    
---------------------
 2761429772394894929
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0', 'city64');
    hash64_string     
----------------------
 -4768102028040737835
(1 row)

select hash64_string('', 'city128');
    hash64_string    
---------------------
 4463240938071824939
(1 row)

select hash64_string('a', 'city128');
    hash64_string     
----------------------
 -3282779286773163312
(1 row)

select hash64_string('abcdefg', 'city128');
    hash64_string    
---------------------
 6752956416143810332
(1 row)

select encode(hash128_string('abcdefg', 'city128'), 'hex');
              encode              
----------------------------------
 1c5ff2ef1452b75d0b491c5915448fc4
(1 row)

select hash64_string('0123456789abcdef', 'city128');
    hash64_string     
----------------------
 -4238272431910926336
(1 row)

select hash64_string('0123456789abcdef0', 'city128');
    hash64_string    
---------------------
 6534406265586176330
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef', 'city128');
    hash64_string    
---------------------
 7077392943806476194
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0', 'city128');
    hash64_string    
---------------------
 8529942145141848512
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcde', 'city128');
    hash64_string     
----------------------
 -5208947204931729004
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef', 'city128');
    hash64_string     
----------------------
 -3020063351221474831
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0', 'city128');
    hash64_string     
----------------------
 -6378252132917199736
(1 row)

select hash64_string('', 'spooky');
    hash64_string    
---------------------
 8819735582944171500
(1 row)

select hash64_string('a', 'spooky');
    hash64_string     
----------------------
 -2619323016173754639
(1 row)

select hash64_string('abcdefg', 'spooky');
    hash64_string    
---------------------
 1147919704542617658
(1 row)

select encode(hash128_string('abcdefg', 'spooky'), 'hex');
              encode              
----------------------------------
 3a20efabe33aee0f6c97f37f73a0ca22
(1 row)

select hash64_string('0123456789abcdef', 'spooky');
    hash64_string    
---------------------
 5147465450635499834
(1 row)

select hash64_string('0123456789abcdef0', 'spooky');
    hash64_string    
---------------------
 6982276862229315490
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef', 'spooky');
    hash64_string     
----------------------
 -6852937307523381414
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0', 'spooky');
    hash64_string    
---------------------
 8060399234958293799
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcde', 'spooky');
    hash64_string    
---------------------
 5567270698085223970
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef', 'spooky');
    hash64_string    
---------------------
 -187398730623730110
(1 row)

select hash64_string('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0', 'spooky');
    hash64_string    
---------------------
 8546626629948030442
(1 row)

SELECT encode(hash128_string('', 'md5'), 'hex');
              encode              
----------------------------------
 d41d8cd98f00b204e9800998ecf8427e
(1 row)

-- d41d8cd98f00b204e9800998ecf8427e
SELECT encode(hash128_string('a', 'md5'), 'hex');
              encode              
----------------------------------
 0cc175b9c0f1b6a831c399e269772661
(1 row)

-- 0cc175b9c0f1b6a831c399e269772661
SELECT encode(hash128_string('abc', 'md5'), 'hex');
              encode              
----------------------------------
 900150983cd24fb0d6963f7d28e17f72
(1 row)

-- 900150983cd24fb0d6963f7d28e17f72
SELECT encode(hash128_string('message digest', 'md5'), 'hex');
              encode              
----------------------------------
 f96b697d7cb7938d525a2f31aaf161d0
(1 row)

-- f96b697d7cb7938d525a2f31aaf161d0
SELECT encode(hash128_string('abcdefghijklmnopqrstuvwxyz', 'md5'), 'hex');
              encode              
----------------------------------
 c3fcd3d76192e4007dfb496cca67e13b
(1 row)

-- c3fcd3d76192e4007dfb496cca67e13b
SELECT encode(hash128_string('ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789', 'md5'), 'hex');
              encode              
----------------------------------
 d174ab98d277d9f5a5611c2c9f419d9f
(1 row)

-- d174ab98d277d9f5a5611c2c9f419d9f
SELECT encode(hash128_string('12345678901234567890123456789012345678901234567890123456789012345678901234567890', 'md5'), 'hex');
              encode              
----------------------------------
 57edf4a22be3c955ac49da2e2107b67a
(1 row)

-- 57edf4a22be3c955ac49da2e2107b67a
--
-- integer hashes
--
select hash_int4(0, 'jenkins');
 hash_int4  
------------
 1800329511
(1 row)

select hash_int4(12345678, 'jenkins');
 hash_int4  
------------
 -824648931
(1 row)

select hash_int4(1234567890123456789::int8, 'jenkins');
 hash_int4  
------------
 1795280340
(1 row)

select hash_int4(0, 'wang32');
 hash_int4  
------------
 -895235421
(1 row)

select hash_int4(12345678, 'wang32');
 hash_int4  
------------
 2029064838
(1 row)

select hash_int4(1234567890123456789::int8, 'wang32');
 hash_int4  
------------
 1072559484
(1 row)

select hash_int4(0, 'wang32mult');
  hash_int4  
-------------
 -1062647446
(1 row)

select hash_int4(12345678, 'wang32mult');
  hash_int4  
-------------
 -1073924033
(1 row)

select hash_int4(1234567890123456789::int8, 'wang32mult');
  hash_int4  
-------------
 -1941958315
(1 row)

select hash_int8(0, 'wang64');
      hash_int8      
---------------------
 8633297058295171728
(1 row)

select hash_int8(12345678, 'wang64');
      hash_int8       
----------------------
 -5256454390535237213
(1 row)

select hash_int8(1234567890123456789::int8, 'wang64');
      hash_int8      
---------------------
 4369460375527774746
(1 row)

select hash_int8(0, 'wang64to32');
 hash_int8 
-----------
 720020139
(1 row)

select hash_int8(12345678, 'wang64to32');
 hash_int8  
------------
 1665754234
(1 row)

select hash_int8(1234567890123456789::int8, 'wang64to32');
 hash_int8 
-----------
 304567056
(1 row)

-- make sure conversion to int4 works correctly
select count(hash_int8(x, 'wang64to32')::int4) from generate_series(1,1000) x;
 count 
-------
  1000
(1 row)

select x, hash_int4(x + 6, 'jenkins') from generate_series(1, 10) x order by hash_int4(x + 6, 'jenkins');
 x  |  hash_int4  
----+-------------
  7 | -1978921976
  6 | -1567819279
  3 | -1246105791
  9 |  -327522260
  1 |  -204548256
 10 |   425555844
  5 |   833732939
  4 |  1712784324
  8 |  1734807528
  2 |  1831203572
(10 rows)

